dbms/src/Common/getNumberOfLogicalCPUCores.cpp

```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <Common/Exception.h>
+#include <Common/getNumberOfLogicalCPUCores.h>
+#include <common/likely.h>
+
+#include <thread>
+
+#if defined(__linux__)
+#include <cmath>
+#include <fstream>
+#endif // __linux__
+
+namespace DB::ErrorCodes
+{
+extern const int CPUID_ERROR;
+} // namespace DB::ErrorCodes
+
+#if defined(__linux__)
+// Try to look at cgroups limit if it is available.
+
+// read int a value from file
+static inline int read_int_from(const char * filename, int default_value)
+{
+    std::ifstream infile(filename);
+    if (!infile.is_open())
+    {
+        return default_value;
+    }
+    int idata;
+    if (infile >> idata)
+    {
+        return idata;
+    }
+    else
+    {
+        return default_value;
+    }
+}
+
+// logical_cpu_cores = min(cpuset.cpus, quota/period)
+static unsigned calCPUCores(int cgroup_quota, int cgroup_period, unsigned cpuset_count)
+{
+    unsigned quota_count = cpuset_count;
+
+    if (cgroup_quota > -1 && cgroup_period > 0)
+    {
+        quota_count = ceil(static_cast<float>(cgroup_quota) / static_cast<float>(cgroup_period));
+    }
+
+    return std::min(cpuset_count, quota_count);
+}
+
+static unsigned read_cpuset_count_from(const char * filename, unsigned default_value)
```

Comment:
Would you like to add some `gtests` to test all of functions？

You can find some example in `man`.

Reply:
I will